### PR TITLE
Fixes/tile renderer enhancements

### PR DESCRIPTION
**Pull request by brendancol**
This branch has a critical tile renderer fix which prevents duplicate tile requests when using multiple tile layers in the same plot.

### CODE REVIEW

**Comment by canavandl** [comment]
Is this supposed to be commented out? :
 
https://github.com/bokeh/bokeh/blob/fixes/tile-renderer-enhancements/bokehjs/src/coffee/renderer/tile/tile_renderer.coffee#L239

I'd remove if unnecessary

**Comment by brendancol** [comment]
Yes, Pruning is not quite ready yet, but close. Will come on a separate PR

**Comment by brendancol** [comment]
I'll remove that now

**Comment by canavandl** [comment]
Is it worth exposing the render timer as a property? 

https://github.com/bokeh/bokeh/blob/fixes/tile-renderer-enhancements/bokehjs/src/coffee/renderer/tile/tile_renderer.coffee#L236

File: bokehjs/src/coffee/renderer/tile/tile_renderer.coffee [hunk 1]
```diff
@@ -85,6 +84,18 @@ class TileRendererView extends PlotWidget
...
+    if @height != @map_frame.get('height') or @width != @map_frame.get('width')
+      extent = @get_extent()
+      zoom_level = @mget('tile_source').get_level_by_extent(extent, @map_frame.get('height'), @map_frame.get('width'))
+      new_extent = @mget('tile_source').snap_to_zoom(extent, @map_frame.get('height'), @map_frame.get('width'), zoom_level)
+      @x_range.set({start:new_extent[0], end: new_extent[2]})
+      @y_range.set({start:new_extent[1], end: new_extent[3]})
+      @extent = new_extent
+      @height = @map_frame.get('height')
+      @width = @map_frame.get('width')
+      return
```
Comment (canavandl) [inline bokehjs/src/coffee/renderer/tile/tile_renderer.coffee]: I don't understand why there's a return statement. Is the intention to stop the render process?

File: bokehjs/src/coffee/renderer/tile/tile_renderer.coffee [hunk 2]
```diff
@@ -85,6 +84,18 @@ class TileRendererView extends PlotWidget
       @_map_data()
       @map_initialized = true
 
+    # brute force way of handling resize or responsive event -------------------------------------------------------------
+    if @height != @map_frame.get('height') or @width != @map_frame.get('width')
```
Comment (canavandl) [inline bokehjs/src/coffee/renderer/tile/tile_renderer.coffee]: It's also a little unclear that ``@height`` and ``@width`` are undefined on the first render, causing the conditional to evaluate as ``True``. Is there a better way to encapsulate this? Perhaps breaking the body into a function and calling it?

File: bokehjs/src/coffee/renderer/tile/tile_renderer.coffee (hunk 1 above)
Comment (brendancol) [inline bokehjs/src/coffee/renderer/tile/tile_renderer.coffee]: Yes. Changing x/y_range will trigger another render which will be needed before loading tiles.

**Comment by canavandl** [comment]
Awesome. I've reviewed the code and tested the examples.

LGTM

**Comment by damianavila** [comment]
@brendancol this is tied to some already opened issue?
Also, I guess #3034 was superseded by this one, but I would like a confirmation (before looking into the code and figured out for myself :wink:)

File: bokeh/models/renderers.py [hunk 3]
```diff
@@ -52,6 +51,10 @@ class TileRenderer(Renderer):
     Specifies the level in which to render the glyph.
     """)
 
+    render_parents = Bool(default=True, help="""
```
Comment (mattpap) [inline bokeh/models/renderers.py]: Will `render_parents` attribute be implemented in other descendants of `Renderer` class, other than `TileRenderer`? If not, then this should be moved to appropriate subclass (`TileRenderer`).

**Comment by mattpap** [comment]
> This branch has a critical tile renderer **fix** (...)

@brendancol, as you used the word **fix**, I would like to see how I can evaluate that this PR is actually a fix? I presume running examples is sufficient. What should I observe? Did you consider adding a test/benchmark for this?

File: bokeh/models/tiles.py [hunk 4]
```diff
@@ -1,7 +1,6 @@
 from __future__ import absolute_import
 from ..plot_object import PlotObject
-
-from ..properties import (Any, Dict, Float, String, Int)
+from ..properties import (Any, Dict, Float, String, Int, Bool)
 
 class TileSource(PlotObject):
```
Comment (mattpap) [inline bokeh/models/tiles.py]: On an unrelated basis, I'm curious if `TileSource` could have anything to do with an actual bokeh data source (`bokeh.models.sources.DataSource`).

File: bokehjs/src/coffee/renderer/tile/tile_renderer.coffee [hunk 5]
```diff
@@ -231,7 +253,8 @@ class TileRenderer extends HasParent
       alpha: 1.0
       x_range_name: "default"
       y_range_name: "default"
-      tile_source: null
+      tile_source: new wmts.Model()
```
Comment (mattpap) [inline bokehjs/src/coffee/renderer/tile/tile_renderer.coffee]: This makes defaults on python and coffeescript side differ.

File: bokeh/models/tiles.py (hunk 4 above)
Comment (brendancol) [inline bokeh/models/tiles.py]: At this moment Tile Source's job is to `manage tile service configuration info` and `construction of tile urls`.  Inheriting from PlotObject, it doesn't have direct relationship to `bokeh.models.sources.DataSource`

File: bokeh/models/renderers.py (hunk 3 above)
Comment (brendancol) [inline bokeh/models/renderers.py]: if Im not mistaken, its already only in the `TileRenderer`.  It is not in `Renderer`

File: bokehjs/src/coffee/renderer/tile/tile_renderer.coffee (hunk 5 above)
Comment (brendancol) [inline bokehjs/src/coffee/renderer/tile/tile_renderer.coffee]: good catch, changing this now.

File: bokeh/models/renderers.py (hunk 3 above)
Comment (mattpap) [inline bokeh/models/renderers.py]: @brendancol, I misread the diff. Everything is fine here.

File: bokeh/models/tiles.py (hunk 4 above)
Comment (brendancol) [inline bokeh/models/tiles.py]: > This branch has a critical tile renderer fix (...)
>>@brendancol, as you used the word fix, I would like to see how I can evaluate that this PR is actually a fix? I presume running examples is sufficient. What should I observe? Did you consider adding a test/benchmark for this?

This fixes the scenario where other bokeh elements call `request_render()`.  This was triggering duplicate requests for tiles pending load. In some cases, tile would be requests 4-5 times.  Simple fix was a `loaded` flag on tiles. 

The specific use case I ran into was with the `map sandwhich` style:
&uarr;`tiled label layer`
&uarr;`dynamic data layer` (e.g. datashader)
&uarr;`tiled base layer`

It was here that demonstrates duplicate tile requests between layers. 
Actually sounds like a decent bokeh example. Show I go ahead and add it?


 ( base layer  data layer /

**Pull request merged**